src/eventloop.c

```diff
@@ -241,45 +161,69 @@ void handle_input(uv_poll_t* handle, int status, int events) {
+    input_event->events = cond;
+    input_event->func = func;
+    input_event->user_data = user_data;
 
-    // Iterate over existing inputs to see if a handle exists.
-    GList* inputListItem = g_list_find_custom(evLoopState.inputs, &fd, findInputByFd);
-    if (inputListItem == NULL) {
-        // We don't have an input handler for that FD yet.
-        poll_handle = malloc(sizeof(uv_poll_t));
-        uv_poll_init(evLoopState.loop, poll_handle, fd);
-        pollsOpen = malloc(sizeof(uint32_t));
-        *pollsOpen = 0;
+    s_evLoopInput *input_handle = g_hash_table_lookup(evLoopState.inputs, &fd);
+    if (input_handle == NULL) {
+        input_handle = g_malloc(sizeof(s_evLoopInput));
+        input_handle->fd = fd;
+        input_handle->handle = g_malloc(sizeof(uv_poll_t));
+        input_handle->cond = cond;
+        input_handle->events = NULL;
+        uv_handle_set_data(input_handle->handle, input_handle);
+        uv_poll_init(evLoopState.loop, input_handle->handle, fd);
+        g_hash_table_insert(evLoopState.inputs, &input_handle->fd, input_handle);
     } else {
-        // We have an existing handle for this, so just take it.
-        s_evLoopInput *previous_handle = (s_evLoopInput*)inputListItem->data;
-        poll_handle = previous_handle->handle;
-        pollsOpen = previous_handle->openPollCounter;
+        // Nothing to do, except update the condition on the poll
+        input_handle->cond = input_handle->cond | cond;
     }
+    // This will update the handle if the cond changed.
+    uv_poll_start(input_handle->handle, input_handle->cond, handle_input);
+    input_event->parent = input_handle;
+    input_handle->events = g_list_append(input_handle->events, input_event);
+    return input_event;
+}
 
-    // Create the input handle.
-    s_evLoopInput *input_handle = malloc(sizeof(s_evLoopInput));
-    input_handle = malloc(sizeof(s_evLoopInput));
-    input_handle->fd = fd;
-    input_handle->handle = poll_handle;
-    input_handle->operations = cond;
-    input_handle->user_data = user_data;
-    input_handle->func = func;
-    input_handle->id = evLoopState.inputId;
-    // Increment the counter because we've added another input to this poll.
-    (*pollsOpen)++;
-    input_handle->openPollCounter = pollsOpen;
-    evLoopState.inputId++;
-
-    // Insert the handler.
-    evLoopState.inputs = g_list_append(evLoopState.inputs, input_handle);
-
-    // enum has 1=readable, 2=writable, which coincidentally matches libpurple's PurpleInputCondition...
-    uv_poll_start(poll_handle, (int)cond, handle_input);
-    return input_handle->id;
+/**
+* Should remove an input handler.  Analogous to g_source_remove in glib.
+* @param handle an identifier, as returned by #input_add.
+* @return       @c TRUE if the input handler was found and removed.
+* @see purple_input_remove
+*/
+gboolean input_remove (guint handle) {
+    g_return_val_if_fail(handle != NULL, false);
+    s_evLoopInputEvent *inputEvent = handle;
```

Comment:
I know that this `guint` is a pointer on both ends, but reading this still gives me chills :D

Maybe it'd be more explicit with `gpointer` instead, and "explicit" conversions using [GPOINTER_TO_UINT](https://developer.gnome.org/glib/stable/glib-Type-Conversion-Macros.html#GPOINTER-TO-UINT:CAPS) and similar?

Reply:
I can do you a `gpointer`, is there a need to do a GPOINTER_TO_UINT?

Reply:
Helps [crossplatformability](https://developer.gnome.org/glib/unstable/glib-Type-Conversion-Macros.html) apparently, and is a little more explicit – which I'd argue C needs a bit more of ;) I'd stick to it just in case, accidentally getting a wrong number into a pointer (when we omit a GUINT_TO_POINTER for instance) could be disastrous when it's actually dereferenced.

Reply:
Had another attempt, does that look better?

Reply:
It does, thanks :)